### PR TITLE
[Docs] Update the conda `environment.yml` to work with Python 3.8

### PR DESCRIPTION
**Pull request by vkorukanti**
## Description
PySpark has dropped support for Python 3.7 which broke the API doc generation script. Update the `docs/environment.yml` dependencies to work with Python 3.8.

## How was this patch tested?
Tested by following the `docs/README.md` and was able to see the generated docs for all modules: delta-spark, flink, connectors and kernel.

### CODE REVIEW

**Comment by felipepessoto** [comment]
@vkorukanti I have this PR to upgrade Python 3.7 -> 3.9 https://github.com/delta-io/delta/pull/1940, but it is already upgraded to 3.8 here: https://github.com/delta-io/delta/commit/4f9c8b9cc294ec7b321847115bf87909c356bc5a

Do you think still make sense to keep, or can I close it?

**Comment by vkorukanti** [comment]
> @vkorukanti I have this PR to upgrade Python 3.7 -> 3.9 #1940, but it is already upgraded to 3.8 here: [4f9c8b9](https://github.com/delta-io/delta/commit/4f9c8b9cc294ec7b321847115bf87909c356bc5a)
> 
> Do you think still make sense to keep, or can I close it?

I think we should keep it? What is the expectation of pyspark when Python 3.8 is used? How will Spark verifies no regressions with 3.8? Sorry for these qns, I am not familiar with the Python ecosystem.

**Comment by felipepessoto** [comment]
Looking it better, Spark supports from 3.8 to 3.11. I don't remember why I thought it was using 3.9 as minimum.
I think we should keep the 3.8, the minimum version. I'll close my PR

File: docs/environment.yml
Comment (allisonport-db) [inline docs/environment.yml:1]: Getting this when creating the environment
```
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - zlib==1.2.13=h5a0b063_0
  - pip==23.2.1=py38hca03da5_0
  - setuptools==68.0.0=py38hca03da5_0
  - wheel==0.41.2=py38hca03da5_0
  - python==3.8.18=hb885b13_0
  - xz==5.4.2=h80987f9_0
  - openssl==3.0.11=h1a28f6b_2
  - readline==8.2=h1a28f6b_0
  - ncurses==6.4=h313beb8_0
  - libcxx==14.0.6=h848a8c0_0
  - sqlite==3.41.2=h80987f9_0
  - tk==8.6.12=hb8d0fd4_0
  - ca-certificates==2023.08.22=hca03da5_0
  - libffi==3.4.4=hca03da5_0
```

I updated the dependencies to be all minor-versions and then it worked
```
name: delta_docs
channels:
  - defaults
dependencies:
  - ca-certificates=2023.08
  - libcxx=14.0
  - libffi=3.4
  - ncurses=6.4
...
```

Comment (vkorukanti) [reply docs/environment.yml:1]: Removed the sha pin. Could you try it again?

**Comment by scottsand-db** [comment]
Has https://docs.conda.io/projects/continuumio-conda/en/latest/user-guide/install/macos.html been moved? 404 page for me

File: docs/README.md [hunk 1]
```diff
@@ -6,7 +6,11 @@ This directory contains scripts to generate API Docs for Scala, Java, and Python
 Follow [Conda Installation Guide](https://docs.conda.io/projects/continuumio-conda/en/latest/user-guide/install/macos.html) to install Anaconda.
```
Comment (scottsand-db) [inline docs/README.md:6]: This link didn't work for me. I just used https://www.anaconda.com/download/

File: docs/README.md [hunk 2]
```diff
@@ -6,7 +6,11 @@ This directory contains scripts to generate API Docs for Scala, Java, and Python
...
 
 ### Create an environment from environment file
-Follow [Create Environment From Environment file](https://docs.conda.io/projects/conda/en/4.6.1/user-guide/tasks/manage-environments.html#create-env-from-file) to create a Conda environment from `/delta/docs/environment.yml` and activate the newly created `docs` environment.
+Follow [Create Environment From Environment file](https://docs.conda.io/projects/conda/en/4.6.1/user-guide/tasks/manage-environments.html#create-env-from-file) to create a Conda environment from `/delta/docs/environment.yml` and activate the newly created `delta_docs` environment.
+
+```
+conda env create --name delta_docs --file=<delta root>/docs/environments.yml
```
Comment (scottsand-db) [inline docs/README.md:12]: `environment.yml` (singular; not plural)

Comment (scottsand-db) [inline docs/README.md:12]: I get error:

```
➜  ~ conda env create --name delta_docs --file=~/delta/docs/environment.yml
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 23.7.4
  latest version: 23.9.0

Please update conda by running

    $ conda update -n base -c defaults conda

Or to minimize the number of packages updated during conda update use

     conda install conda=23.9.0



Downloading and Extracting Packages

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Installing pip dependencies: failed

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda/exception_handler.py", line 17, in __call__
        return func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda_env/cli/main.py", line 56, in do_call
        exit_code = getattr(module, func_name)(arguments, parser)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda/notices/core.py", line 124, in wrapper
        return func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda_env/cli/main_create.py", line 162, in execute
        result[installer_type] = installer.install(
                                 ^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda_env/installers/pip.py", line 76, in install
        return _pip_install_via_requirements(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda_env/installers/pip.py", line 42, in _pip_install_via_requirements
        requirements = Utf8NamedTemporaryFile(
                       ^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/site-packages/conda/auxlib/compat.py", line 38, in Utf8NamedTemporaryFile
        return NamedTemporaryFile(
               ^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/tempfile.py", line 563, in NamedTemporaryFile
        file = _io.open(dir, mode, buffering=buffering,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/tempfile.py", line 560, in opener
        fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/scott.sandre/anaconda3/lib/python3.11/tempfile.py", line 256, in _mkstemp_inner
        fd = _os.open(file, flags, 0o600)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    FileNotFoundError: [Errno 2] No such file or directory: '/Users/scott.sandre/~/delta/docs/condaenv.qiicu26t.requirements.txt'

`$ /Users/scott.sandre/anaconda3/bin/conda-env create --name delta_docs --file=~/delta/docs/environment.yml`

  environment variables:
                 CIO_TEST=<not set>
  CONDA_AUTO_UPDATE_CONDA=false
                CONDA_EXE=/Users/scott.sandre/anaconda3/bin/conda
         CONDA_PYTHON_EXE=/Users/scott.sandre/anaconda3/bin/python
               CONDA_ROOT=/Users/scott.sandre/anaconda3
              CONDA_SHLVL=0
           CURL_CA_BUNDLE=<not set>
                 INFOPATH=/usr/local/share/info:
               LD_PRELOAD=<not set>
                  MANPATH=/usr/local/share/man::
                     PATH=/Users/scott.sandre/anaconda3/condabin:/usr/local/opt/python@3.7/bin:/
                          Users/scott.sandre/docs/bin:/usr/local/bin:/usr/local/sbin:/usr/local/
                          bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/h
                          omebrew/bin:/opt/homebrew/sbin:/var/run/com.apple.security.cryptexd/co
                          dex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptex
                          d/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/
                          codex.system/bootstrap/usr/appleinternal/bin:/Users/scott.sandre/Libra
                          ry/Python/3.9/bin
       REQUESTS_CA_BUNDLE=<not set>
            SSL_CERT_FILE=<not set>

     active environment : None
            shell level : 0
       user config file : /Users/scott.sandre/.condarc
 populated config files : /Users/scott.sandre/.condarc
          conda version : 23.7.4
    conda-build version : 3.26.1
         python version : 3.11.5.final.0
       virtual packages : __archspec=1=arm64
                          __osx=13.5.1=0
                          __unix=0=0
       base environment : /Users/scott.sandre/anaconda3  (writable)
      conda av data dir : /Users/scott.sandre/anaconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-arm64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-arm64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/scott.sandre/anaconda3/pkgs
                          /Users/scott.sandre/.conda/pkgs
       envs directories : /Users/scott.sandre/anaconda3/envs
                          /Users/scott.sandre/.conda/envs
               platform : osx-arm64
             user-agent : conda/23.7.4 requests/2.31.0 CPython/3.11.5 Darwin/22.6.0 OSX/13.5.1 aau/0.4.2 c/q3zhavAK8lAz687x_JwChg s/eBpXRReed_rPJ8wjRe9dsA e/Ov9RKLgFae4liphkhRX-zQ
                UID:GID : 502:20
             netrc file : None
           offline mode : False


An unexpected error has occurred. Conda has prepared the above report.
If you suspect this error is being caused by a malfunctioning plugin,
consider using the --no-plugins option to turn off plugins.

Example: conda --no-plugins install <package>

Alternatively, you can set the CONDA_NO_PLUGINS environment variable on
the command line to run the command without plugins enabled.

Example: CONDA_NO_PLUGINS=true conda install <package>
```

File: docs/environment.yml
Comment (allisonport-db) [reply docs/environment.yml:1]: Works for me now

File: docs/README.md (hunk 2 above)
Comment (allisonport-db) [reply docs/README.md:12]: Just checked and I got this error when running with absolute path for --file but it works for me with a relative path

Comment (allisonport-db) [reply docs/README.md:12]: ¯\_(ツ)_/¯

Comment (allisonport-db) [reply docs/README.md:12]: `conda env create --name delta_docs --file=docs/environment.yml` works for me

Comment (scottsand-db) [reply docs/README.md:12]: > Just checked and I got this error when running with absolute path for --file but it works for me with a relative path

Yikes that's crazy. Using absolute worked for me. Maybe let's just mentioned to use one of)
1. absolute
2. relative to ~
3. relative to <delta home>/docs folder

**Review by scottsand-db — APPROVED** [review]

File: docs/README.md (hunk 1 above)
Comment (vkorukanti) [reply docs/README.md]: updated the link

File: docs/README.md (hunk 2 above)
Comment (vkorukanti) [reply docs/README.md]: Added the comment to explicit mentioning absolute path is needed.

**Pull request closed without merging**